### ISSUE TITLE
Access Denied [Error 403] Simple sharepoint list query 

### ISSUE THREAD

**Issue by russgove** [pnp/pnpjs#1718]
### Category
- [ ] Enhancement
- [ ] Bug
- [X] Question
- [ ] Documentation gap/issue

### Version
THIS IS RELATED TO ISSUE 1000: https://github.com/pnp/pnpjs/issues/1000

Please specify what version of the library you are using: [ 2.3.0 ]

Please specify what version(s) of SharePoint you are targeting: [SPO ]

*If you are not using the latest release, please update and see if the issue is resolved before submitting an issue.*

### Expected / Desired Behavior / Question
I have the following code in the reneder method of a domain isolated webpart.
```
  public render(): void {
    var x = Web('https://russellwgove.sharepoint.com/sites/testazad/').lists.getById('a2141c79-39df-49be-868e-4715d6a4562a')
      .get()
      .then((x) => {
        debugger;
      })
      .catch((err) => {
        debugger;
      });
```
I expect to get the list back
### Observed Behavior
When deployed , the call returns an http 403 eror. 
When I run in the workbench it works fine

doing a trace on the failed call i see that it is not sending the token across when run in a domain isolated webpart:
```
GET https://russellwgove.sharepoint.com/sites/testazad/_api/web/lists('a2141c79-39df-49be-868e-4715d6a4562a') HTTP/1.1
Host: russellwgove.sharepoint.com
Connection: keep-alive
Cache-Control: max-age=0
sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="90", "Google Chrome";v="90"
accept: application/json
x-clientservice-clienttag: PnPCoreJS:2.3.0:ls.getById
sec-ch-ua-mobile: ?0
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.85 Safari/537.36
content-type: application/json;odata=verbose;charset=utf-8
Origin: https://russellwgove-appa762c11aaa3d4013b30bcc4c0386f290.sharepoint.com
Sec-Fetch-Site: same-site
Sec-Fetch-Mode: cors
Sec-Fetch-Dest: empty
Referer: https://russellwgove-appa762c11aaa3d4013b30bcc4c0386f290.sharepoint.com/
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9,zh;q=0.8,fr-FR;q=0.7,fr;q=0.6,nl;q=0.5
```

### Steps to Reproduce
Add the above code to the onrender method of a domain isolated webpart

**Comment by patrick-rodgers** [comment]
@russgove - haven't had a chance to work on recreating this one yet. My guess is there is a trick to work in domain isolated webparts and we haven't documented it. Will try and have something by the next release if I can find the time. Thanks!

**Comment by russgove** [comment]
Thanks Patrick!

**Comment by juliemturner** [comment]
@russgove I can shed some light on the issue, but I'm not sure what your goal is by using Web({sharepoint url}) so I'm not sure if it's going to help or not, but fundamentally your issue is that you're trying to make a request to a site you do not have access to. 

If the SharePoint URL you're using in Web is theoretically supposed to be the same web as the page where the web part is hosted, then the URL you have is not the correct one.  If you make the call like this instead:

`var x = sp.web.lists.getById('a2141c79-39df-49be-868e-4715d6a4562a').get().then()....` 

then that will work for you.

If you then check out the network tab in your browser, you're going to find that the URL that the call is coming from is the isolated URL that was created for your isolated web part which is then running in an iFrame on the page.  So yeah, it's `isolated` by definition from the rest of SharePoint.

Long story short, you will not be able to directly call the web, or any other web for that matter.  If that's your goal you're going to either have to turn off web part isolation or build an API that will act as an intermediary to your web part to get the data you need in other sites. If that's not your goal, then I think you just need to adjust how you're making your calls to use the pattern I show above.

**Comment by russgove** [comment]
Thanks Julie! I turned off isolation and its been working fine so this is
not  a priority for me right now, but would like to add it back sometime.

I have a webpart that needs to be able to create a list in any site in my
tenant, so i figured that it kinda needs to be isolated. Are you saying I
cant do that in an isolated webpart, maybe move my code to an azure
function?

**Comment by juliemturner** [comment]
To be honest that's really not what isolated web parts are made for, they're really about autorizing a web part to make calls to either an API that you created or the Microost Graph without allowing other web parts to do so.  This would be used for things like, you want this web part to be able to access the users one drive for them, but you don't want any other SPFx solution to be able to access users one drive.  Keep in mind client-side solutions use "delegated" permissions by default, which means whomever is authenticated can only do what they actually have permissions to do. What you're describing is really elevated permissions.  You want that web part to be able to create this list regardless of whomever is running it, which means that you really can't use SharePoint API to do it directly from the client. Yes, Azure Function is one way to do those elevated privileges... that you would make sure that the web part is only being called by an authenticated user and then the azure function would use an AAD App Registration to elevate permissions to create the list. I have an old but somewhat relevant blog post on the subject, might get you started but it's out of date and is calling the function from a Flow. https://julieturner.net/2017/07/sharepoint-flow-azure-functions-launching-a-microsoft-flow-from-client-side-code/

**Comment by PathToSharePoint** [comment]
> I have a webpart that needs to be able to create a list in any site in my tenant

Wouldn't the [Graph API for SharePoint](https://docs.microsoft.com/en-us/graph/api/list-create?view=graph-rest-beta&tabs=http) work in this scenario? I haven't tried, but apparently it can run under application permissions.

**Comment by russgove** [comment]
Ah, right. I think i misstated my intention above (serves me right for
replying on a friday evening :-))

The app IS calling a custom api , THATS why I made it an isolated webpart.
I was relying on the fact the the users were sharepoint admins so that they
would have permission to create the list. So in an isolated webpart an
admin could not connect to the other web.

**Comment by juliemturner** [comment]
@PathToSharePoint  - No. You cannot make a call to a Graph API and expect to use "app permissions" from the web client, it's only delegated permissions unless you build your own API that then uses app permissions for you. You want to validate that the user that's making the request of your API has permissions to access your API (via delegated permissions).. but you can't use app level permissions directly.

@russgove  - Then that is the correct use of an isolated web part. Your web part has permissions to your API and that's it so your API needs to do the work of creating the list for them, you can't create the list from the client and then call your API, your API needs to do all the work. The problem is your URL for that web parts site is like https://{abc123}.sharepoint.com vs https://mytenant.sharepoint.com and therefore is domain isolated, you do not have a valid token to call any of the SharePoint API's for any site outside that domain isolated site, ergo why Web("{some other site collection}") doesn't work.

**Comment by patrick-rodgers** [comment]
Closing this issue as answered. If you have additional questions or we did not answer your question, please open a new issue, ref this issue, and provide any additional details available. Thank you!
